### PR TITLE
feat(lyrics-plus): add japanese lyrics conversion tool

### PR DESCRIPTION
**Pull request by 41pha1**
This adds the functionality of switching between different display modes for Japanese lyrics.

### CODE REVIEW

**Comment by theRealPadster** [comment]
Probably should have the dictionaries included in the same repo

**Comment by 41pha1** [comment]
> Probably should have the dictionaries included in the same repo

Alright I included them in the PR now

**Comment by CharlieS1103** [comment]
@theRealPadster You sure? Seems like it would add quite a bit of bloat to the install, although I didn't really check.

**Comment by 41pha1** [comment]
Loading them dynamically using jsdeliver seems to work just fine

File: CustomApps/lyrics-plus/index.js
Comment (ririxiqt) [inline CustomApps/lyrics-plus/index.js:520]: shouldn't it be
```suggestion
                    lyrics: CONFIG.visual["translate"] && translatedLyrics ? translatedLyrics : this.state.unsynced,
```
since it's about unsyced lyrics, not synced ones

File: CustomApps/lyrics-plus/manifest.json
Comment (ririxiqt) [inline CustomApps/lyrics-plus/manifest.json:82]: well... i would rather load these dict files when it's needed. 30mb is a lot, since it adds another megabytes to download for people, even when they don't want it

**Review by ririxiqt — CHANGES_REQUESTED** [review]

**Comment by ririxiqt** [comment]
lmao I can't type xD ^

File: CustomApps/lyrics-plus/manifest.json
Comment (41pha1) [reply CustomApps/lyrics-plus/manifest.json:82]: Yes that's quite understandable, how about making it so that there's a download button when first opening the conversion menu on a Japanese song? The dicts will be cached locally after that.

File: CustomApps/lyrics-plus/README.md
Comment (kyrie25) [inline CustomApps/lyrics-plus/README.md:25]:
```suggestion
Choose between different option of displaying Japanese lyrics. (Furigana, Romaji, Hirgana, Katakana)
```

File: CustomApps/lyrics-plus/README.md
Comment (kyrie25) [inline CustomApps/lyrics-plus/README.md:42]:
```suggestion
-   The algorithm for converting Japanese lyrics is based on [Hexenq's Kuroshiro](https://github.com/hexenq/kuroshiro).
```

File: CustomApps/lyrics-plus/index.js
Comment (kyrie25) [inline CustomApps/lyrics-plus/index.js:230]: What does this one-liner do exactly? Did you mean to assign both of them the value `null`?

**Review by kyrie25 — CHANGES_REQUESTED** [review]
This is really amazing work! 😄 Thanks a lot for your contribution.
I just have a few small suggestions regarding this PR.

File: CustomApps/lyrics-plus/manifest.json
Comment (kyrie25) [reply CustomApps/lyrics-plus/manifest.json:82]: Since both the dictionary files are already inside the custom app folder, this means that it will be included in CLI releases and users would only need to download them whenever they upgrade the CLI. Sounds reasonable to me tbh, but it still means that the CLI package is now 10 times larger than it ever was:
image

If we don't want to do this, then the only option left is to *not* include it into CLI releases and use it as a remote script via `jsdelivr` like before. This would mean that users can prompt when to download the dictionary, but the client would need to do it every time cache is dumped.

Comment (CharlieS1103) [reply CustomApps/lyrics-plus/manifest.json:82]: To be entirely honest I think this is the best solution, multiplying the cli size by 10 does not seem worth it to me. Particularly since a majority of people probably won't ever do so much as touch this feature. Or, perhaps we could merge the cli changes but make the modified lyrics plus it's own separate repo. (Kind of similar to when we made it so custom app owners could run code like an extension for marketplace, instead of building marketplace into the cli repo like expFeatures or home/sidebar config.)

Comment (CharlieS1103) [reply CustomApps/lyrics-plus/manifest.json:82]: Ah, just checked back and saw that there are actually no cli changes, why not just make this a separate repo altogether?

**Comment by CharlieS1103** [comment]
How about we make it so lyrics plus uses his translation or code or whatever initially with just hotloading but if the user installs the dicts locally via an install script then it uses that method instead? So users who don't need the feature don't need the extra 30mb and the users who often use the feature and don't want the load times can install it

File: CustomApps/lyrics-plus/index.js
Comment (41pha1) [reply CustomApps/lyrics-plus/index.js:230]:
```suggestion
        this.state.furigana = this.state.romaji = this.state.hirgana = this.state.katakana = null;
```

Comment (kyrie25) [reply CustomApps/lyrics-plus/index.js:230]: Yep that sounds about right, commit whenever you see fit.

**Comment by 41pha1** [comment]
The dictionaries will now be dynamically loaded if they are not yet present and the conversion menu will show a loading indicator. 

If the dicts are somehow added to the client in any other way, it will use them instead and skip the downloading step.

(I havent come up with a better place than my [private repo](https://github.com/41pha1/spicetify-romaji-lyrics/tree/main/marketplace) yet)

File: CustomApps/lyrics-plus/Translator.js
Comment (41pha1) [inline CustomApps/lyrics-plus/Translator.js:6051]: This should probably change

File: CustomApps/lyrics-plus/index.js
Comment (kyrie25) [inline CustomApps/lyrics-plus/index.js:281]:
```suggestion
        let lyricsToTranslate = this.state.synced;

        if (!lyricsToTranslate) {
        	lyricsToTranslate = this.state.unsynced;
        	return;
        }
```
Also is the variable reassigned anywhere? If not then use `const` instead of `let`.

Comment (ririxidev) [reply CustomApps/lyrics-plus/index.js:281]: well, no it doesn't make sense, it should be
```suggestion
        let lyricsToTranslate = this.state.synced ?? this.state.unsynced;

        if (!lyricsToTranslate) return;
```

Comment (41pha1) [reply CustomApps/lyrics-plus/index.js:281]: > well, no it doesn't make sense, it should be

Yea this seems to clear it up a bit!

File: CustomApps/lyrics-plus/Utils.js
Comment (ririxidev) [inline CustomApps/lyrics-plus/Utils.js:73]:
```suggestion
        const rubyElems = s.split("<ruby>");
        const reactChildren = [];
```

File: CustomApps/lyrics-plus/index.js
Comment (ririxidev) [inline CustomApps/lyrics-plus/index.js:292]:
```suggestion
                const translatedLines = result.split("\n");
```

File: CustomApps/lyrics-plus/index.js
Comment (ririxidev) [inline CustomApps/lyrics-plus/index.js:500]:
```suggestion
        const translatedLyrics = this.state[CONFIG.visual["translation-mode"]];
```

File: CustomApps/lyrics-plus/index.js
Comment (ririxidev) [inline CustomApps/lyrics-plus/index.js:561]:
```suggestion
        const showTranslationButton = (this.state.synced || this.state.unsynced) && Utils.isJapanese(this.state.synced || this.state.unsynced) && (mode == SYNCED || mode == UNSYNCED);
        const translatorLoaded = this.translator.finished;
```

**Review by ririxidev — CHANGES_REQUESTED** [review]
Please run https://prettier.io  on all javascript files as well

File: CustomApps/lyrics-plus/Utils.js
Comment (ririxidev) [inline CustomApps/lyrics-plus/Utils.js:79]:
```suggestion
            const kanji = rubyElems[i].split("<rp>")[0];
            const furigana = rubyElems[i].split("<rt>")[1].split("</rt>")[0];
```

I didn't notice it at first glance and these variables aren't re-assigned so they can be assigned as `const`

**Review by ririxidev — CHANGES_REQUESTED** [review]
That's all I noticed.
About dict files, we'll see if we'll keep them in different repository instead of your own

File: CustomApps/lyrics-plus/Utils.js
Comment (41pha1) [reply CustomApps/lyrics-plus/Utils.js:79]: > That's all I noticed. About dict files, we'll see if we'll keep them in different repository instead of your own

Thats sound just about perfect!  Thanks for all of the suggestions, ill make sure to use const more often in the future xD

**Review by ririxidev — COMMENTED** [review]
Please resolve conflicts that appeared after merging #1992. We created an repo for dict files - https://github.com/spicetify/spicetify-lyrics-romaji. Is it okay with you if we will keep these dict files in this repo instead?

**Comment by 41pha1** [comment]
> Please resolve conflicts that appeared after merging #1992. We created an repo for dict files - https://github.com/spicetify/spicetify-lyrics-romaji. Is it okay with you if we will keep these dict files in this repo instead?

That seems like the perfect place for them! Could you add the dict files? I sadly cant seem to open a PR on an empty repo.
Also prettier seems to be erroring after merging, looks like I might have messed something up?

**Comment by ririxidev** [comment]
Don't worry about Prettier Action. For some reason it doesn't use your fork but instead this repo.
I will upload these files when I will be able to and will let you know

**Comment by ririxidev** [comment]
#2000 probably fixed prettier action in this pull request. Additionally, I uploaded dictionary files to [spicetify-lyrics-romaji](https://github.com/spicetify/spicetify-lyrics-romaji) repo which you can find [here](https://github.com/spicetify/spicetify-lyrics-romaji/tree/main/dictionary)

**Comment by 41pha1** [comment]
> #2000 probably fixed prettier action in this pull request. Additionally, I uploaded dictionary files to [spicetify-lyrics-romaji](https://github.com/spicetify/spicetify-lyrics-romaji) repo which you can find [here](https://github.com/spicetify/spicetify-lyrics-romaji/tree/main/dictionary)

Awesome thanks ^^ I changed the urls to now point to the new location and it seems to work!

**Review by ririxidev — APPROVED** [review]
i need to look into prettier action again :/

**Review by kyrie25 — APPROVED** [review]
LGTM

**Comment by huhridge** [comment]
I tested this, and it works great! I was gonna add this, but this is better than I would have implemented it.

Is there a possibility of adding a check to only translate when the song is actually Japanese? Right now, it passes everything, and it messes with the formatting of the lyrics.

**Comment by 41pha1** [comment]
> I tested this, and it works great! I was gonna add this, but this is better than I would have implemented it.
> 
> Is there a possibility of adding a check to only translate when the song is actually Japanese? Right now, it passes everything, and it messes with the formatting of the lyrics.

Yes that should have been implemented, I also made a function for it but somehow never called it? Thanks for pointing this out!

**Comment by huhridge** [comment]
> > I tested this, and it works great! I was gonna add this, but this is better than I would have implemented it.
> > Is there a possibility of adding a check to only translate when the song is actually Japanese? Right now, it passes everything, and it messes with the formatting of the lyrics.
> 
> Yes that should have been implemented, I also made a function for it but somehow never called it? Thanks for pointing this out!

Just tried it, and works perfectly now! You did call the function once to check whether to display the translation menu iirc.

**Comment by 41pha1** [comment]
> prettify your code with prettier but with config from this repo 🙄

Looks like its all working now :D

**Comment by ririxidev** [comment]
Wouldn't it be better to include https://github.com/hexenq/kuroshiro js files in our repo and then invoke stuff from Translator.js after loading them by jsdeliver?

**Comment by 41pha1** [comment]
> Wouldn't it be better to include https://github.com/hexenq/kuroshiro js files in our repo and then invoke stuff from Translator.js after loading them by jsdeliver?

Probably however I think it would be difficult to achieve, as this also uses the analyzer functions here https://github.com/hexenq/kuroshiro-analyzer-kuromoji
and I have rewritten a decent amount of code that somehow seemed to be broken in the original repo (I think due to JavaScript version changes, I am not quite sure. Even the demo page at https://kuroshiro.org/ seems to be broken now) 
as well as added support for loading the dictionaries dynamically.

**Comment by ririxidev** [comment]
Could you provide source code for this then which I will upload to repo?

**Comment by 41pha1** [comment]
Do you mean the code of the modified kuroshiro / analyzer bundle that currently sits along the Translator class in Translator.js?
I sadly don't have separated files anywhere as I exported everything directly to the bundle with npm.

I am not quite sure if I understand correctly :/

**Comment by ririxidev** [comment]
So, you don't have any source-code for this code? Why's that? We will not be able to update a code then. We can't extract source from webpack compiled code. If you deleted it from disk you should be able to restore it with DMDE

**Comment by 41pha1** [comment]
Oh yea my bad! I added the unminified code and split it into kuromoji.js and kuroshiro.js If wanted we can move them to the separate repo?

**Comment by ririxidev** [comment]
Revert this commit please. I didn't mean unminified code but source code of it. This is still bundled by webpack which is not a source code

**Comment by 41pha1** [comment]
I am afraid I cant provide you with what you are looking for in that case

**Comment by kyrie25** [comment]
> I am afraid I cant provide you with what you are looking for in that case

We are asking if you can provide for us the repository/package in which you forked those files. If those codes are not owned by you, legally we cannot use it as we may risk breaking the terms of the license in which the code is published under.

**Comment by 41pha1** [comment]
> We are asking for the original code in which this was compiled to, if that clarifies anything for you.
> 
> Currently what you have committed here is the minified/compiled code from a Webpack build process, of which we cannot inspect, verify and maintain its functionality.

I will have to get to work on the files from the Kuroshiro repo again, because I was as naive as to modify the compiled code from npm directly and thus am not and never haven been in the possession of the uncompiled code.

**Comment by huhridge** [comment]
Is there a possibility of bundling the translate function within Spicetify? So maybe it can be used in other extensions and custom apps, etc?

**Comment by kyrie25** [comment]
> Is there a possibility of bundling the translate function within Spicetify? So maybe it can be used in other extensions and custom apps, etc?

The extension will be released as a standalone package so feel free to use that instead.

**Comment by huhridge** [comment]
Yeah, but the problem is importing modules in the extensions and making it work with marketplace, unless that was fixed?

**Comment by kyrie25** [comment]
> Yeah, but the problem is importing modules in the extensions and making it work with marketplace, unless that was fixed?

That would be discussed on the Marketplace side of development. You can always choose to opt out on having users install it via Marketplace and do it locally instead. We are also planning to make an option for developers to redirect users to their repository's Readme instead of installing by default (spicetify/spicetify-marketplace#346)
Either way, we are not forcing users to increase their installation package size by ten-fold for a feature that has probably around 1-2% of the userbase using it.

**Comment by Pithaya** [comment]
In case it helps, I've also been trying to integrate Kuroshiro in a spicetify extension without having to download the dictionary files locally, and got it to work without having to host the files in a repo. You can directly get the dictionary files from npm.
There is an [issue](https://github.com/takuyaa/kuromoji.js/issues/37) with loading external urls however (I guess this is why the library's code had to be edited here to use downloaded dictionary files directly?), so I had to add a fix (found [here](https://github.com/mobilusoss/textlint-browser-runner/pull/7/files)) to get it working.

The repo is private for now so I can't link the file but here is the code I'm using (the extension is built with spicetify creator):
```ts
import Kuroshiro from 'kuroshiro';
import KuromojiAnalyzer from '../../node_modules/kuroshiro-analyzer-kuromoji/dist/kuroshiro-analyzer-kuromoji.min.js';

type ExtendedXMLHttpRequest = XMLHttpRequest & {
    patched_open: (method: string, url: string) => void;
};

const dictPath = 'https://cdn.jsdelivr.net/npm/kuromoji@0.1.2/dict';
const dictPathWithoutDoubleSlash =
    'https:/cdn.jsdelivr.net/npm/kuromoji@0.1.2/dict';
/**
 * Fix an issue with kuromoji when loading dict from external urls
 * See: https://github.com/takuyaa/kuromoji.js/issues/37
 * Adapted from: https://github.com/mobilusoss/textlint-browser-runner/pull/7
 */
export function applyKuromojiFix(): void {
    (XMLHttpRequest.prototype as ExtendedXMLHttpRequest).patched_open =
        XMLHttpRequest.prototype.open;
    XMLHttpRequest.prototype.open = function (method: string, url: string) {
        if (url.indexOf(dictPathWithoutDoubleSlash) === 0) {
            (this as ExtendedXMLHttpRequest).patched_open(
                method,
                url.replace(dictPathWithoutDoubleSlash, dictPath)
            );
        } else {
            (this as ExtendedXMLHttpRequest).patched_open(method, url);
        }
    };
}

export async function createKuroshiro(): Promise<Kuroshiro> {
    applyKuromojiFix();

    const kuroshiro: Kuroshiro = new Kuroshiro();
    const analyzer: KuromojiAnalyzer = new KuromojiAnalyzer({
        dictPath: dictPath,
    });

    await kuroshiro.init(analyzer);

    return kuroshiro;
}

```

**Comment by Overrevvv** [comment]
Thank you for your contribution and this fantastic extension. I'd like to know if we can convert Romanji further into English.

**Comment by ririxidev** [comment]
@Pithaya if you want to integrate this with lyrics-plus, please create a pull request. This one is on hold because pr's author can't provide source code with his changes - and since your code can get dictionary files from jsdelivr, it's even better for us 👍

**Comment by ohitstom** [comment]
i really hope this isnt abandoned lol

**Comment by 41pha1** [comment]
I think I got it to work by loading kuroshiro, kuromoji and the dictionary files via JSDeliver directly from npm using the XMLHttp fix similar to the one suggested by [Pithaya](https://github.com/Pithaya).

**Comment by ririxidev** [comment]
It's okay by me, @41pha1

**Review by kyrie25 — APPROVED** [review]
LGTM

**Review by ririxidev — APPROVED** [review]

**Comment by Tetrax-10** [comment]
@41pha1 

In line [114](https://github.com/spicetify/spicetify-cli/blob/master/CustomApps/lyrics-plus/OptionsMenu.js#L114) should that be `isJapanese` or `!isJapanese` Because if its a japanese song then the menu should show up right?

image

**Pull request merged**
